### ISSUE TITLE
aws_cloudfront_distribution attribute mismatch: diffs didn't  match during apply.

### ISSUE THREAD

**Issue by spikelampson** [terraform-providers/terraform-provider-aws#3256]
```
An execution plan has been generated and is shown below.                                  
Resource actions are indicated with the following symbols:
  + create                                              
  ~ update in-place
                             
Terraform will perform the following actions:               
                                                                          
  ~ aws_cloudfront_distribution.static-facebook                                                          
      origin.2528787734.custom_header.#:        "0" => "0"                                               
      origin.2528787734.custom_origin_config.#: "0" => "0"                                               
      origin.2528787734.domain_name:            "facebook.babylonhealth.com.s3.amazonaws.com" => ""      
      origin.2528787734.origin_id:              "facebook.babylonhealth.com" => ""                       
      origin.2528787734.origin_path:            "" => ""                                                 
      origin.2528787734.s3_origin_config.#:     "0" => "0"                                               
      origin.3355674526.custom_header.#:        "0" => "0"                                               
      origin.3355674526.custom_origin_config.#: "0" => "0"                                               
      origin.3355674526.domain_name:            "" => "facebook.babylonhealth.com"                       
      origin.3355674526.origin_id:              "" => "facebook.babylonhealth.com"                       
      origin.3355674526.origin_path:            "" => ""                                                 
      origin.3355674526.s3_origin_config.#:     "0" => "0"                                               
                                                                                                         
  + aws_route53_record.facebook-static-prod                                                              
      id:                                       <computed>                                               
      fqdn:                                     <computed>                                               
      name:                                     "facebook.babylonhealth.com"                             
      records.#:                                "1"                                                      
      records.2327122678:                       "drd5xh28nit1u.cloudfront.net"                           
      ttl:                                      "30"                                                     
      type:                                     "CNAME"                                                  
      zone_id:                                  "Z2ULNG6JQKRJR0"                                         
                                                                                                         
  ~ aws_s3_bucket.facebook-babylonhealth                                                                 
      website_domain:                           "s3-website.eu-west-2.amazonaws.com" => "facebook.babylon
health.com"                                                                                              


Plan: 1 to add, 2 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes
                          
1 error(s) occurred:                                                                                     
                                                                                                         
* aws_cloudfront_distribution.static-facebook: aws_cloudfront_distribution.static-facebook: diffs didn't 
match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.               
                                                                                                         
Please include the following information in your report:                                                 
                          
    Terraform Version: 0.11.3                       
    Resource ID: aws_cloudfront_distribution.static-facebook                                             
    Mismatch reason: attribute mismatch: origin.3355674526.custom_header.#                               
            Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"origin.2528787734.s3_origin_config.#":*terraform.ResourceAttrDiff{O
ld:"0", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sens
itive:false, Type:0x0}, "origin.2528787734.custom_header.#":*terraform.ResourceAttrDiff{Old:"0", New:"0",
 NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Typ
e:0x0}, "origin.3355674526.domain_name":*terraform.ResourceAttrDiff{Old:"", New:"facebook.babylonhealth.c
om", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false,
 Type:0x0}, "origin.2528787734.custom_origin_config.#":*terraform.ResourceAttrDiff{Old:"0", New:"0", NewC
omputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0
}, "origin.3355674526.custom_header.#":*terraform.ResourceAttrDiff{Old:"0", New:"0", NewComputed:false, N
ewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.33556
74526.origin_path":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:false, NewRemoved:false, NewEx
tra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.3355674526.s3_origin_config
.#":*terraform.ResourceAttrDiff{Old:"0", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interface
 {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.3355674526.custom_origin_config.#":*terr
aform.ResourceAttrDiff{Old:"0", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil),
 RequiresNew:false, Sensitive:false, Type:0x0}, "origin.3355674526.origin_id":*terraform.ResourceAttrDiff
{Old:"", New:"facebook.babylonhealth.com", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil
), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.2528787734.domain_name":*terraform.ResourceAttr
Diff{Old:"facebook.babylonhealth.com.s3.amazonaws.com", New:"", NewComputed:false, NewRemoved:true, NewEx
tra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.2528787734.origin_id":*terr
aform.ResourceAttrDiff{Old:"facebook.babylonhealth.com", New:"", NewComputed:false, NewRemoved:true, NewE
xtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.2528787734.origin_path":*t
erraform.ResourceAttrDiff{Old:"", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface {}(nil),
 RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:fals
e, Meta:map[string]interface {}(nil)}
              Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:m
ap[string]*terraform.ResourceAttrDiff{"origin.2528787734.custom_header.#":*terraform.ResourceAttrDiff{Old
:"0", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensit
ive:false, Type:0x0}, "origin.1376845673.origin_path":*terraform.ResourceAttrDiff{Old:"", New:"", NewComp
uted:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, 
"origin.2528787734.custom_origin_config.#":*terraform.ResourceAttrDiff{Old:"0", New:"0", NewComputed:fals
e, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.2
528787734.origin_path":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:false, NewRemoved:true, Ne
wExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.1376845673.s3_origin_con
fig.#":*terraform.ResourceAttrDiff{Old:"0", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interf
ace {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.2528787734.domain_name":*terraform.Re
sourceAttrDiff{Old:"facebook.babylonhealth.com.s3.amazonaws.com", New:"", NewComputed:false, NewRemoved:t
rue, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.2528787734.s3_ori
gin_config.#":*terraform.ResourceAttrDiff{Old:"0", New:"0", NewComputed:false, NewRemoved:false, NewExtra
:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.1376845673.origin_id":*terrafo
rm.ResourceAttrDiff{Old:"", New:"facebook.babylonhealth.com", NewComputed:false, NewRemoved:false, NewExt
ra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.1376845673.custom_header.#":*terraform.ResourceAttrDiff{Old:"0", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.1376845673.custom_origin_config.#":*terraform.ResourceAttrDiff{Old:"0", New:"0", NewComputed:false, NewRemoved:false, NewExtra:interface{}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.1376845673.domain_name":*terraform.ResourceAttrDiff{Old:"", New:"s3-website.eu-west-2.amazonaws.com", NewComputed:false, NewRemoved:false, NewExtra:interface{}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "origin.2528787734.origin_id":*terraform.ResourceAttrDiff{Old:"facebook.babylonhealth.com", New:"", NewComputed:false, NewRemoved:true, NewExtra:interface{}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}                    
```

### Expected Behavior
Plan should be applied

### Actual Behavior
Result shown above. Terraform believes that it is out of sync with AWS, but everything seems to be working as expected.

### Steps to Reproduce
1. `terraform apply`

**Comment by sppwf** [comment]
Same Issue for me happened :(

**Comment by slonokot** [comment]
same here

**Comment by slonokot** [comment]
it looks like the issue resolved with upgrading of aws provider to 11.4

**Comment by bflad** [comment]
Hi folks 👋 Sorry for the trouble with the `aws_cloudfront_distribution` resource. Since its original implementation, there were many configuration blocks that did not benefit from certain difference simplifications and arguments that were incorrectly using ordered lists instead of sets. Usually these type of implementation details surface in unexpected ways during Terraform runs as `diffs didn't match during apply` error in Terraform 0.11 and prior. (Aside: in Terraform 0.12 and later, these types of errors will include much better diagnostic information rather than dumping the raw differences.)

The configuration block simplifications and fixes were applied to the resource logic in version 2.0.0 of the Terraform AWS Provider, which will be releasing very shortly. Existing configurations should mostly work as they did previously, but `aws_cloudfront_distribution` resource updates and recreations should perform more reliably now without these errors. Since there were a large amount of varying `diffs didn't match during apply` reports, we are opting to close all these issues which appear to fall under this category.

We encourage everyone to file a new GitHub issue once upgraded to version 2.0.0 of the Terraform AWS Provider, should there be any further problems, so we can appropriately triage them. Thanks for your understanding and hope the upgrade is helpful. 😄
